File: src/main/java/icu/samnyan/aqua/sega/maimai2/handler/impl/UploadUserPortraitHandler.java
package icu.samnyan.aqua.sega.maimai2.handler.impl;

import com.fasterxml.jackson.core.JsonProcessingException;
import icu.samnyan.aqua.sega.maimai2.handler.BaseHandler;
import icu.samnyan.aqua.sega.maimai2.model.request.UploadUserPortrait;
import icu.samnyan.aqua.sega.util.jackson.BasicMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.*;
import java.util.Base64;
import java.util.Map;

/**
 * @author samnyan (privateamusement@protonmail.com)
 */
@Component("Maimai2UploadUserPortraitHandler")
public class UploadUserPortraitHandler implements BaseHandler {

    private static final Logger logger = LoggerFactory.getLogger(UploadUserPortraitHandler.class);

    private final BasicMapper mapper;

    private final String picSavePath;
    private final boolean enable;
    private final long divMaxLength;

    public UploadUserPortraitHandler(BasicMapper mapper,
                                     @Value("${game.maimai2.userPhoto.enable:true}") boolean enable,
                                     @Value("${game.maimai2.userPhoto.picSavePath:data/userPhoto}") String picSavePath,
                                     @Value("${game.maimai2.userPhoto.divMaxLength:16}") long divMaxLength) {
        this.mapper = mapper;
        this.picSavePath = picSavePath;
        this.enable = enable;
        this.divMaxLength = divMaxLength;

        if (enable) {
            try {
                Files.createDirectories(Paths.get(picSavePath));
            } catch (Exception ignored) {
            }
        }
    }

    @Override
    public String handle(Map<String, Object> request) throws JsonProcessingException {
        /*
        Maimai DX sends splited base64 data for one jpeg image.
        So, make a temp file and keep append bytes until last part received.
        If finished, rename it to other name so user can keep save multiple score cards in a single day.
        */

        if (enable) {
            var uploadUserPhoto = mapper.convert(request, UploadUserPortrait.class);
            var userPhoto = uploadUserPhoto.getUserPortrait();

            long userId = userPhoto.getUserId();
            int divNumber = userPhoto.getDivNumber();
            int divLength = userPhoto.getDivLength();
            String divData = userPhoto.getDivData();

            if (divLength > divMaxLength)
                return "{\"returnCode\":-1,\"apiName\":\"com.sega.maimai2servlet.api.UploadUserPortraitApi\"}";

            try {
                var tmp_filename = Paths.get(picSavePath, userId + "-up.tmp");
                if (divNumber == 0)
                    Files.deleteIfExists(tmp_filename);

                byte[] imageData = Base64.getDecoder().decode(divData);
                Files.write(tmp_filename, imageData, StandardOpenOption.CREATE, StandardOpenOption.APPEND);

                logger.info(String.format("received user %d photo data %d/%d", userId, divNumber + 1, divLength));

                if (divNumber == (divLength - 1)) {
                    var filename = Paths.get(picSavePath, userId + "-up.jpg");
                    Files.move(tmp_filename, filename, StandardCopyOption.REPLACE_EXISTING);

                    userPhoto.setDivData("");
                    var userPortaitMetaJson = mapper.write(userPhoto);
                    var json_filename = Paths.get(picSavePath, userId + "-up.json");
                    Files.write(json_filename, userPortaitMetaJson.getBytes(StandardCharsets.UTF_8), StandardOpenOption.CREATE, StandardOpenOption.WRITE, StandardOpenOption.TRUNCATE_EXISTING);

                    logger.info(String.format("saved user %d photo data", userId));
                }
            } catch (IOException e) {
                logger.error("Result: User photo save failed", e);
            }
        }
        return "{\"returnCode\":1,\"apiName\":\"com.sega.maimai2servlet.api.UploadUserPortraitApi\"}";
    }
}

File: src/main/java/icu/samnyan/aqua/sega/maimai2/handler/UploadUserPortraitHandler.kt
package icu.samnyan.aqua.sega.maimai2.handler

import ext.div
import ext.path
import icu.samnyan.aqua.net.utils.PathProps
import icu.samnyan.aqua.sega.general.BaseHandler
import icu.samnyan.aqua.sega.maimai2.model.request.UploadUserPortrait
import icu.samnyan.aqua.sega.util.jackson.BasicMapper
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.io.IOException
import java.nio.file.Files
import java.nio.file.StandardCopyOption.REPLACE_EXISTING
import java.nio.file.StandardOpenOption.APPEND
import java.nio.file.StandardOpenOption.CREATE
import java.util.*
import kotlin.io.path.writeText

@Component("Maimai2UploadUserPortraitHandler")
class UploadUserPortraitHandler(
    val mapper: BasicMapper,
    paths: PathProps,
    @param:Value("\${game.maimai2.userPhoto.enable:true}") val enable: Boolean,
    @param:Value("\${game.maimai2.userPhoto.divMaxLength:32}") val divMaxLength: Long
) : BaseHandler {
    val path = paths.mai2Portrait.path()
    val success = """{"returnCode":1,"apiName":"com.sega.maimai2servlet.api.UploadUserPortraitApi"}"""
    val fail = """{"returnCode":-1,"apiName":"com.sega.maimai2servlet.api.UploadUserPortraitApi"}"""

    override fun handle(request: Map<String, Any>): String {
        if (!enable) return success

        // Maimai DX sends split base64 data for one jpeg image.
        // So, make a temp file and keep append bytes until last part received.
        // If finished, rename it to other name so user can keep save multiple scorecards in a single day.

        val up = mapper.convert(request, UploadUserPortrait::class.java).userPortrait

        val id = up.userId
        val num = up.divNumber
        val len = up.divLength

        if (len > divMaxLength) {
            logger.warn("stop user $id uploading photo data because divLength($len) > divMaxLength($divMaxLength)")
            return fail
        }

        try {
            val tmp = path / "$id-up.tmp"
            if (num == 0) Files.deleteIfExists(tmp)
            Files.write(tmp, Base64.getDecoder().decode(up.divData), CREATE, APPEND)

            logger.info("> User photo $id data ${num + 1}/${len}")

            if (num == (len - 1)) {
                Files.move(tmp, path / "$id-up.jpg", REPLACE_EXISTING)

                up.divData = ""
                (path / "$id-up.json").writeText(mapper.write(up))

                logger.info("> Saved user $id photo data")
            }
        } catch (e: IOException) {
            logger.error("> User photo save failed", e)
        }

        return success
    }

    companion object {
        private val logger: Logger = LoggerFactory.getLogger(UploadUserPortraitHandler::class.java)
    }
}
